Use new UEFI bootloader

tsoutsman opened the pull request:
The bootloader is at <https://github.com/theseus-os/uefi-bootloader>. It will also need to be reviewed.

Depends on #785, #786, and #790. I haven't been able to properly review all the changes in this PR because it builds on top of many different PRs. Once they are merged, I'll be able to review the changes properly.

tsoutsman committed:
Update rust toolchain

Older Rust toolchains didn't have the `rust-std` component for
`*-unknown-uefi` targets meaning they couldn't be used for artifact
dependencies.

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Update `uefi_builder`

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Implement new bootloader

Works with tsoutsman/uefi-bootloader#9c85f8e.

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Fix build for BIOS

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Dirty

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Add `BootInformation::additional_reserved_memory_regions`

This is the first step in supporting Theseus loaded at an aribtrary
address.

The changes to `frame_allocator` are necessary for the frame allocator
to correctly initialise when a reserved memory region overlaps with more
than one reserved memory region that came before it. Reshuffling the
order that the bootloader info and modules memory regions are added to
the reserved memory regions revealed the problem.

The PR also removes `bios` as the default feature for `nano_core` as it
makes it impossible to compile for UEFI, as discussed on Discord.

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Add `nano_core/bios` feature to Clippy run

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Don't reuse recursive entry in `PageTable::with`

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Add current mapper to `PageTable::with` closure

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Use current page table to calculate stack and boot info frames

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Cleanup

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Remove kernel and user stack references

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Update docs

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Merge branch 'boot-spec-additional-memory-regions' into arbitrary-load-address

tsoutsman committed:
Merge branch 'access-current-page-table-2' into arbitrary-load-address

tsoutsman committed:
Calculate end of deginated low pages via translation

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Translate section virtual addresses in `find_section_memory_bounds`

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Use physical addresses for identity mapping nano core sections

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Merge branch 'arbitrary-load-address' into new-uefi-bootloader

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

kevinaboos committed:
Clarify docs

tsoutsman committed:
Add default nano core feature

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Refactor `kernel_memory_region`

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Update kernel/kernel_config/src/memory.rs

Co-authored-by: Kevin Boos <1139460+kevinaboos@users.noreply.github.com>

tsoutsman committed:
Fix `TEMP_P4` address

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Update `PageTable::with` docs

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Update temporary recursive index naming

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Merge branch 'access-current-page-table-2' into new-uefi-bootloader

tsoutsman committed:
Merge branch 'boot-spec-additional-memory-regions' into new-uefi-bootloader

tsoutsman committed:
Dirty

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Merge branch 'theseus_main' into new-uefi-bootloader

tsoutsman committed:
Working commit on `x86_64`

Syncs with tsoutsman/uefi-bootloader#32b8dfd.

tsoutsman committed:
Add new `uefi_builder`

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Change `uefi-bootloader` path dependencies to git dependencies

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Remove `aarch64` changes

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman commented on kernel/nano_core/src/asm/defines.asm:
```diff
@@ -16,6 +16,7 @@ AP_NMI_LINT         equ TRAMPOLINE + 56
 AP_NMI_FLAGS        equ TRAMPOLINE + 64
 AP_MAX_FB_WIDTH     equ TRAMPOLINE + 72
 AP_MAX_FB_HEIGHT    equ TRAMPOLINE + 80
+AP_GDT				equ TRAMPOLINE + 88
```
This shows up differently in my editor.

kevinaboos replied:
looks like you've used tabs instead of spaces

tsoutsman committed:
More cleanup

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

kevinaboos commented:
Thanks, I've reviewed all the PR this depends on and am ready to merge them pending some quick discussions.

tsoutsman committed:
Update `Cargo.lock`

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Merge branch 'theseus_main' into new-uefi-bootloader-no-aarch

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Remove changes to `rustfmt.toml`

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Merge branch 'theseus_main' into new-uefi-bootloader-no-aarch

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Remove changes to `vga_buffer`

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Final cleanup

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Update Rust toolchain

Older Rust toolchains didn't have the `rust-std` component for
`*-unknown-uefi` targets meaning they couldn't be used for artifact
dependencies.

The upcoming switch to the new UEFI bootloader uses artifact
dependencies on `*-unknown-uefi` targets.

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Clarify GDT entries

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Merge branch 'theseus_main' into new-uefi-bootloader-no-aarch

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Update version

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Satisfy Clippy

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Satisfy Clippy

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Merge branch 'new-rust-version' into new-uefi-bootloader-no-aarch

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman commented:
Depends on #803.

kevinaboos commented on kernel/boot_info/src/lib.rs:
```diff
@@ -122,7 +122,7 @@ pub trait BootInformation: 'static {
 
     /// Returns the end of the kernel's image in physical memory.
-    fn kernel_end(&self) -> Result<PhysicalAddress, &'static str>;
+    fn kernel_end(&self) -> Result<VirtualAddress, &'static str>;
```
fix doc comment above

kevinaboos commented on kernel/boot_info/src/lib.rs:
```diff
@@ -38,7 +38,7 @@ pub trait ElfSection {
     fn start(&self) -> VirtualAddress;
 
-    /// Returns the section's length.
+    /// Returns the section's length in memory.
```
is this comment addition intended to contrast against the section's length in the ELF file? if so, let's say that too.

kevinaboos commented on kernel/nano_core/src/lib.rs:
```diff
@@ -194,10 +214,10 @@ where
         log::info!("The nano_core (in loadable mode) is invoking the captain init function: {:?}", section.name);
 
-        type CaptainInitFunc = fn(MmiRef, NoDrop<EarlyIdentityMappedPages>, NoDrop<Stack>, VirtualAddress, VirtualAddress, Option<PhysicalAddress>) -> Result<(), &'static str>;
+        type CaptainInitFunc = fn(MmiRef, NoDrop<EarlyIdentityMappedPages>, NoDrop<Stack>, VirtualAddress, VirtualAddress, VirtualAddress, Option<PhysicalAddress>) -> Result<(), &'static str>;
```
according to your other docs, the ap_gdt is a physical address

kevinaboos commented on kernel/memory/src/lib.rs:
```diff
@@ -220,7 +221,15 @@ pub fn init(
+    page_allocator::init(
+        VirtualAddress::new(
+            Mapper::from_current()
+                .translate(boot_info.kernel_end()?)
```
is this right? Looking at the calculation of `kernel_end`, it seems like an exclusive upper bound. I didn't dig into it in detail, but what happens if `kernel_end()` is aligned to the start of the next page, causing this `translate()` action to fail?

kevinaboos replied:
~~ok on second inspection, this is a bit odd. The boot_info provides the kernel end vaddr, which we then translate to physical, which we the use to obtain the identity vaddr? I legit cannot tell if that's equivalent to the previous code, lol~~
EDIT: nvm, i'm tired lol. The answer is clearly that the boot loader isn't guaranteed to have identity mapped it.

kevinaboos commented on tools/uefi_builder/common/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+name = "uefi_builder_common"
+version = "0.1.0"
+edition = "2021"
+authors = ["Klim Tsoutsman <klim@tsoutsman.com>"]
```
please add a description (and in other Cargo.toml files too)

kevinaboos reviewed (commented):
Minor to-do: add a description of the `uefi_builder*` tool crates to the `tools/README.md` file

kevinaboos commented on kernel/memory/src/paging/mapper.rs:
```diff
@@ -50,6 +50,9 @@ use owned_borrowed_trait::{OwnedOrBorrowed, Owned, Borrowed};
 /// that it is only invoked for `UnmappedFrames`.
 pub(super) static INTO_ALLOCATED_FRAMES_FUNC: Once<fn(FrameRange) -> AllocatedFrames> = Once::new();
 
+pub fn translate(virtual_address: VirtualAddress) -> Option<PhysicalAddress> {
```
```suggestion
/// A convenience function to translate the given virtual address into a
/// physical address using the currently-active page table.
pub fn translate(virtual_address: VirtualAddress) -> Option<PhysicalAddress> {
```

tsoutsman committed:
Minor fixes

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

tsoutsman committed:
Update kernel/memory/src/paging/mapper.rs

Co-authored-by: Kevin Boos <1139460+kevinaboos@users.noreply.github.com>

tsoutsman committed:
Add crate descriptions

Signed-off-by: Klim Tsoutsman <klim@tsoutsman.com>

kevinaboos committed:
simplify/clarify

kevinaboos committed:
placate clippy

kevinaboos commented:
LGTM; will approve & merge pending a review of `uefi-bootloader`

kevinaboos committed:
Merge branch 'theseus_main' into new-uefi-bootloader-no-aarch

kevinaboos committed:
update to latest uefi-bootloader version

kevinaboos committed:
ensure same version of uefi-bootloader* crates are used
* top-level Cargo.lock file must match lockfiles in `tools/uefi-builder/*`
